File: Go/JumpGameIII.go
//go:build ignore

package main

import "fmt"

func canReach(arr []int, start int) bool {
	n := len(arr)
	seen := make([]bool, n)
	queue := []int{start}
	seen[start] = true
	for len(queue) > 0 {
		i := queue[0]
		queue = queue[1:]
		if arr[i] == 0 {
			return true
		}
		for _, j := range []int{i + arr[i], i - arr[i]} {
			if j >= 0 && j < n && !seen[j] {
				seen[j] = true
				queue = append(queue, j)
			}
		}
	}
	return false
}

func main() {
	fmt.Println(canReach([]int{4, 2, 3, 0, 3, 1, 2}, 5)) // true
	fmt.Println(canReach([]int{4, 2, 3, 0, 3, 1, 2}, 0)) // true
	fmt.Println(canReach([]int{3, 0, 2, 1, 2}, 2))       // false
}

File: CPP/JumpGameIII.cpp
#include <iostream>
#include <queue>
#include <vector>

class Solution {
public:
    bool canReach(std::vector<int>& arr, int start) {
        int n = arr.size();
        std::vector<bool> seen(n, false);
        std::queue<int> q;
        q.push(start);
        seen[start] = true;
        while (!q.empty()) {
            int i = q.front();
            q.pop();
            if (arr[i] == 0) return true;
            for (int j : {i + arr[i], i - arr[i]}) {
                if (j >= 0 && j < n && !seen[j]) {
                    seen[j] = true;
                    q.push(j);
                }
            }
        }
        return false;
    }
};

int main() {
    Solution s;
    std::vector<int> a1 = {4, 2, 3, 0, 3, 1, 2};
    std::vector<int> a2 = {3, 0, 2, 1, 2};
    std::cout << std::boolalpha;
    std::cout << s.canReach(a1, 5) << "\n"; // true
    std::cout << s.canReach(a1, 0) << "\n"; // true
    std::cout << s.canReach(a2, 2) << "\n"; // false
    return 0;
}
